agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/HttpClientMethodVisitor.java

```diff
@@ -102,6 +104,25 @@ public void onMethodEnter() {
         mv.visitMethodInsn(INVOKEINTERFACE, "org/apache/http/RequestLine", "getUri", "()Ljava/lang/String;", true);
         uriLocal = this.newLocal(Type.getType(Object.class));
         mv.visitVarInsn(ASTORE, uriLocal);
+
+        //Load HttpUriRequest instance into Local Array. Contains information of Path and Host
+        mv.visitVarInsn(ALOAD, 2);
+        mv.visitMethodInsn(INVOKEINTERFACE, "org/apache/http/client/methods/HttpUriRequest", "getURI", "()Ljava/net/URI;", true);
```

Comment:
@dhaval24 - have you tested this? You're relying on org.apache.http.client.methods.HttpUriRequest being present at index 2, but the method this is intrumenting has org.apache.http.HttpRequest in the signature. I'm not sure if it is a problem, but it might be. In addition, the URI was already being collected (in line 104), so you can just use that in the CoreAgentNotificationHandler to extract the host and path (instead of changing this visitor). That would be safer and cleaner.

Reply:
@grlima yes I have tested this out. You can see the test results in the screen shot of this PR. Now coming to your point on that instrumenting method has signature HttpRequest I think instead instrumenting method has signature as follows `public CloseableHttpResponse execute(HttpUriRequest request) throws IOException, ClientProtocolException`. Here is the link to docs : https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/impl/client/CloseableHttpClient.html#execute(org.apache.http.client.methods.HttpUriRequest) 
```
CloseableHttpClient client = HttpClients.createDefault();
HttpGet get = new HttpGet("https://www.abc.com");
    try {
      HttpResponse response = client.execute(get);
      System.out.println(response.getStatusLine().getStatusCode());
      client.close();
    }
```

In the above execute method calls the method mentioned above. So in that case I think I am instrumenting the right method with right signature.

There is already a url variable which can be reused? 
A. I think answer here is no. Because `getURI()` returns a string and doesn't have required components like Absolute Path and target (aka host). here is the reference to the docs (http://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/methods/HttpUriRequest.html#getURI())

Now coming to the question that we should instead parse the URI string to get the desired target, relative path etc. I think I would refrain from doing these because of following reason:
String manipulation with regex is expensive operation, also it is susceptible to errors. Further URLs may not be that simple as that appears : for example = https:///www.xyz.com:8888/name=Dhaval?....... so on a long one. In that case coming up with a exact regex which will efficiently exactly the required host and target is challenging.

I think instrumentation here is pretty straight forward and I do not see any possibility of exceptions which could arrise. For a matter of fact I have tried this on SpringBoot, Spring and J2EE applications and things seemed fine. Let me know if you have further questions on this.

Reply:
What about `java.net.URL`? It's supposed to handle absolute and relative URLs and provide you an ability to get any portion of URL back as a string, e.g. hostname, query string and etc..
Having one standard parsing instead of byte-code instrumentation sounds like a safer way.

Reply:
@Dmitry-Matveev as I mentioned above flow goes as follows...
client.execute(get) -> public CloseableHttpResponse execute(HttpUriRequest request, HttpContext context). It is this class of ApacheHttpClient which holds the java.net.URI class. Inside the URI you can get almost anything that you want, as a string (hostname, query, path, raw url etc.). So ideally we only need to instrument the code to capture this URI object which sits inside the HttpUriRequest and that is what my code does. I did not purposefully touch the other URL string coming from org.apache.http.RequestLine to avoid introducing any possible bugs, but that refactoring is very well possible and I do not see any harm in triming that out.

Also we could use the same class HttpUriRequest to get the method (GET/ POST/PUT) instead of instrumenting RequestLine for consistency. Does this make sense? I can explain more if needed.

Reply:
I meant that if you have a string of absolute or relative URL already (like @grlima suggested in his review), then you can avoid additional byte-code instrumentation and just create new URL object from that string to get host and path.

Reply:
@Dmitry-Matveev I think we can keep single instrumentation here. I will remove the instrumentation for RequestLine method which got only String URL but instead keep the instrumentation for java.net.URL here which can provide us all the required information. I will make those refactoring.